### ISSUE TITLE
Improve memory usage with Hadoop Streaming

### ISSUE THREAD

**Issue by whitfin** [whitfin/vessel#5]
Erlang is eager with `stdio` which makes it very memory intensive on Hadoop Streaming, as it reads the entire input into memory without a buffer, and before it's requested. This means that your files have to fit into memory, which is sometimes unrealistic. 

This cannot be fixed inside the Vessel application code, but perhaps we could ship a binary wrapper which pipes through the input at a rate that doesn't bloat the Erlang memory usage. 

This is labelled an enhancement as it's an Erlang attribute, rather than a Vessel bug. It may be that we cannot solve this effectively (in which case, just tune as mentioned in the README).

**Comment by Trevoke** [comment]
Could we do something like working with streams in Erlang? https://github.com/tim/erlang-streams or https://github.com/rpt/estreams or https://github.com/epappas/erl_streams 

None of them have examples involving loading files, unfortunately.

Elixir has Streams in the stdlib and it's very convenient. Maybe that's something to explore? https://hexdocs.pm/elixir/Stream.html

**Comment by zackehh** [comment]
@Trevoke aloha, but unfortunately it's an issue in the BEAM rather than the application layer.

Vessel already uses Streams (see [here](https://github.com/zackehh/vessel/blob/master/lib/vessel/pipe.ex#L22-L26)), but the issue is that the BEAM loads `:stdio` eagerly so even though you're streaming `:stdin`, it still reads the entire thing into memory on startup.

You can try that yourself by piping large IO into an Elixir archive which just sleeps (or whatever); it'll eagerly load the buffer even though you're not actually asking for it, or even interacting with it at all.

That's why this is such an annoyance to fix; you would need some sort of binding around Vessel in another language which operates as a buffer (and even then, there has to be message passing between BEAM and the binding as it can't just pipe straight through due to the same IO loading).

I was going to roll something in Rust or C which you can just request the next chunk of input with, but never got around to it as it feels like something which should stand as a separate project.

Does all of this make sense?

If you have any other ideas/suggestions/solutions/questions/etc, feel free to drop comments in here; I'm a bit lost as the best way to solve this issue, but it does need fixing at some point.

**Comment by Trevoke** [comment]
In Elixir, we can stream files (tested with a 1.9GB file, unless that loaded to memory in under 5 seconds).

Here is a simple example from the [documentation](https://hexdocs.pm/elixir/Stream.html#run/1) of loading a file and making a change to the contents and streaming it back out to another file without loading it all into memory:

```elixir
stream = File.stream!("code")
|> Stream.map(&String.replace(&1, "#", "%"))
|> Stream.into(File.stream!("new"))
|> Stream.run
```


Down in Erlang, we can stream down at this level

- http://erlang.org/doc/man/file.html#read-2
- http://erlang.org/doc/man/file.html#pread-2

For compression, apparently `file:open/2` has a `compressed` option: http://erlang.org/doc/man/file.html#open-2

-----

So... I don't understand why we need to be working with `:stdio` ?

**Comment by zackehh** [comment]
@Trevoke 

You question tells me that you're unfamiliar with the Hadoop Streaming protocol? The idea behind the protocol is that everything is done via `:stdio` in order to avoid rolling your own file handling (and basically rewriting Hadoop itself). We're tied to `:stdio` due to the protocol; using files directly is not an option.

Thanks for the info, but I am already very familiar with Streams - as previously mentioned Vessel *already uses them*. Streaming `:stdio` is very different to streaming a file in that the BEAM eagerly loads `:stdin` regardless of whether you're ready to use it (basically making the streaming aspect useless). This is a BEAM-ism rather than application logic behaving badly; you can see the issue by doing literally nothing in your application but piping files in and watching your memory go up and up and up.

Does that make sense?

**Comment by Trevoke** [comment]
Yeah.. It makes perfect sense. I thought we were implementing more of the
mangling process, which is why I figured the stream might make sense, I
didn't realize we were just picking them up when Hadoop gives them to us,
but that makes sense too.
Sorry about hassling you :)

**Comment by zackehh** [comment]
@Trevoke no problemo! I think maybe I assumed that knowledge in the original issue, so my apologies if it was unclear :)
